opentelemetry-stdout/src/metrics/exporter.rs

```diff
@@ -76,57 +76,139 @@ impl PushMetricsExporter for MetricsExporter {
+                // TODO
+            } else if let Some(sum) = data.downcast_ref::<data::Sum<u64>>() {
+                println!("\t\tType         : Sum");
+                print_sum(sum);
+            } else if let Some(sum) = data.downcast_ref::<data::Sum<i64>>() {
+                println!("\t\tType         : Sum");
+                print_sum(sum);
+            } else if let Some(sum) = data.downcast_ref::<data::Sum<f64>>() {
+                println!("\t\tType         : Sum");
+                print_sum(sum);
+            } else if let Some(gauge) = data.downcast_ref::<data::Gauge<u64>>() {
+                println!("\t\tType         : Gauge");
+                print_gauge(gauge);
+            } else if let Some(gauge) = data.downcast_ref::<data::Gauge<i64>>() {
+                println!("\t\tType         : Gauge");
+                print_gauge(gauge);
+            } else if let Some(gauge) = data.downcast_ref::<data::Gauge<f64>>() {
+                println!("\t\tType         : Gauge");
+                print_gauge(gauge);
+            } else {
+                println!("Unsupported data type");
+            }
+        });
+    }
+}
+
+fn print_sum<T: std::fmt::Debug + DataPointValue>(sum: &data::Sum<T>) {
+    println!("\t\tSum DataPoints");
+    println!("\t\tMonotonic    : {}", sum.is_monotonic);
+    print_data_points(&sum.data_points);
+}
+
+fn print_gauge<T: std::fmt::Debug + DataPointValue>(gauge: &data::Gauge<T>) {
+    println!("\t\tGauge DataPoints");
+    print_data_points(&gauge.data_points);
+}
+
+fn print_histogram<T: std::fmt::Debug + DataPointValue>(histogram: &data::Histogram<T>) {
+    println!("\t\tHistogram DataPoints");
+    print_hist_data_points(&histogram.data_points);
+}
+
+fn print_data_points<T: std::fmt::Debug + DataPointValue>(data_points: &[data::DataPoint<T>]) {
+    for (i, data_point) in data_points.iter().enumerate() {
+        println!("\t\tDataPoint #{}", i);
+        println!("\t\t\tValue        : {:#?}", data_point.value);
+        println!("\t\t\tAttributes   :");
+        for kv in data_point.attributes.iter() {
+            println!("\t\t\t\t ->  {}: {}", kv.key, kv.value.as_str());
+        }
+    }
+}
+
+fn print_hist_data_points<T: std::fmt::Debug + DataPointValue>(
+    data_points: &[data::HistogramDataPoint<T>],
+) {
+    for (i, data_point) in data_points.iter().enumerate() {
+        println!("\t\tDataPoint #{}", i);
+        println!("\t\t\tCount        : {}", data_point.count);
+        println!("\t\t\tSum          : {:?}", data_point.sum);
+        if let Some(min) = &data_point.min {
+            println!("\t\t\tMin          : {:?}", min);
+        }
+
+        if let Some(max) = &data_point.max {
+            println!("\t\t\tMax          : {:?}", max);
+        }
+
+        println!("\t\t\tAttributes   :");
+        for kv in data_point.attributes.iter() {
+            println!("\t\t\t\t ->  {}: {}", kv.key, kv.value.as_str());
+        }
+    }
+}
+
+trait DataPointValue {}
```

Comment:
Why is this trait needed?

Reply:
was a way to ensure <T> (metric value) can be printed for f64,i64,u64

Reply:
It should be able to print that without using this trait.

Reply:
T has no bounds specified! The <T> in the data structure exposed to exporter is totally unrelated to the <T> in hot path..I think we can fix this when we refactor exporter/collect part of metrics

Reply:
I don't follow. Does this trait have anything to do with collect?

> T has no bounds specified! 

Why do we need this additional bound? If you simply remove this trait `DataPointValue` and update the methods that you have added in the PR to only have the `Default` bound, everything should work just fine.

```diff
- fn print_sum<T: std::fmt::Debug + DataPointValue>(sum: &data::Sum<T>)
+ fn print_sum<T: std::fmt::Debug>(sum: &data::Sum<T>)

// Same thing for `print_gauge`, `print_histogram`, `print_data_points`, `print_hist_data_points`
```

If it works without this additional trait bound, what value is this new trait adding?

Reply:
Removed that unnecessary trait. (I am not sure why I even added that earlier 🤣 )